File: ts-sdk/client/src/txbuilder/closeMarket.ts
import { Address, GetAccountInfoApi, GetMultipleAccountsApi, IInstruction, Rpc, TransactionSigner } from "@solana/kit";
import { getDeactivateLookupTableInstruction } from "@solana-program/address-lookup-table";

import { fetchMarket, getCloseMarketInstruction, getMarketAddress, getTunaConfigAddress } from "../index.ts";

export async function closeMarketInstructions(
  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi>,
  authority: TransactionSigner,
  poolAddress: Address,
  deactivateLookupTable: boolean,
): Promise<IInstruction[]> {
  const tunaConfig = (await getTunaConfigAddress())[0];
  const marketAddress = (await getMarketAddress(poolAddress))[0];
  const market = await fetchMarket(rpc, marketAddress);

  const instructions: IInstruction[] = [];

  if (deactivateLookupTable) {
    instructions.push(
      getDeactivateLookupTableInstruction({
        authority,
        address: market.data.addressLookupTable,
      }),
    );
  }

  instructions.push(
    getCloseMarketInstruction({
      authority,
      tunaConfig,
      market: marketAddress,
      vaultA: market.data.vaultA,
      vaultB: market.data.vaultB,
    }),
  );

  return instructions;
}

File: rust-sdk/client/src/txbuilder/close_market.rs
use crate::accounts::fetch_market;
use crate::instructions::CloseMarket;
use crate::{get_market_address, get_tuna_config_address};
use anyhow::Result;
use solana_client::rpc_client::RpcClient;
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;
use spl_token::solana_program::address_lookup_table::instruction::deactivate_lookup_table;

pub fn close_market_instruction(
    rpc: &RpcClient,
    authority: &Pubkey,
    pool: &Pubkey,
    deactivate_address_lookup_table: bool,
) -> Result<Vec<Instruction>> {
    let tuna_config_address = get_tuna_config_address().0;
    let market_address = get_market_address(pool).0;
    let market = fetch_market(&rpc, &market_address)?;

    let mut instructions = vec![];

    if deactivate_address_lookup_table {
        instructions.push(deactivate_lookup_table(market.data.address_lookup_table, *authority));
    }

    let ix_builder = CloseMarket {
        authority: *authority,
        tuna_config: tuna_config_address,
        market: market_address,
        vault_a: market.data.vault_a,
        vault_b: market.data.vault_b,
    };

    instructions.push(ix_builder.instruction());

    Ok(instructions)
}
